### PR TITLE
Add dropdown for selecting fullscreen resolution

### LINKED ISSUE #2706: Can’t choose custom fullscreen resolution
Currently, you can not use any other resolution other than native (no resolution select in the settings menu). Reason of issue: low performance due to peoples PC's not being able to handle everything ofc.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/59128074-can-t-choose-custom-fullscreen-resolution?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UselessToucan**
resolves #2706 
depends on https://github.com/ppy/osu-framework/pull/1589

https://youtu.be/xf5E9GEzLHQ

- [x] ppy/osu-framework#1860

### CODE REVIEW

**Comment by FreezyLemon** [comment]
That would of course be ok 👍 .
Just be aware that the PR might still change in some regards.

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs [hunk 1]
```diff
@@ -62,6 +78,15 @@ private void load(FrameworkConfigManager config)
                 },
             };
 
+            windowModeDropdown.Bindable.ValueChanged += (s) =>
```
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: replace `(s)` with `_`

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs [hunk 2]
```diff
@@ -72,5 +97,12 @@ private void load(FrameworkConfigManager config)
             };
             letterboxing.TriggerChange();
         }
+
+        private IEnumerable<KeyValuePair<string, int>> getResolutions()
```
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: why make this a function? can probably just be inline.

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs [hunk 3]
```diff
@@ -72,5 +97,12 @@ private void load(FrameworkConfigManager config)
...
+        {
+            var availableDisplayResolutions = (game.Window as DesktopGameWindow)?.AvailableDisplayResolutions;
+
+            return (availableDisplayResolutions ?? throw new InvalidOperationException()).Select((t, i) => new KeyValuePair<string, int>($"{t.Width}x{t.Height}@{t.RefreshRate}Hz", i));
```
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: why throw an `InvalidOperationException`? what happens in headless execution?

**Review by peppy — DISMISSED** [review]

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs [hunk 4]
```diff
@@ -19,11 +19,13 @@ public class LayoutSettings : SettingsSubsection
         private FillFlowContainer letterboxSettings;
 
         private Bindable<bool> letterboxing;
+        private Bindable<Size> sizeFullscreen;
```
Comment (FreezyLemon) [inline osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: Any reason for using `Bindable<Size>` here over `BindableSize`?

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs [hunk 5]
```diff
@@ -32,6 +34,17 @@ private void load(FrameworkConfigManager config, OsuGame game)
             this.game = game;
 
             letterboxing = config.GetBindable<bool>(FrameworkSetting.Letterboxing);
+            sizeFullscreen = config.GetBindable<Size>(FrameworkSetting.SizeFullscreen);
+
+            var resolutions = getResolutions();
+            var resolutionDropdownBindable = new BindableInt(resolutions.FirstOrDefault(r => r.Key.StartsWith($"{sizeFullscreen.Value.Width}x{sizeFullscreen.Value.Height}")).Value);
+
+            resolutionDropdownBindable.ValueChanged += _ =>
```
Comment (FreezyLemon) [inline osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: You should change the parameter name from `_` to something else.
The osu project normally only uses `_` as a name for unused parameters in lambdas.

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs (hunk 4 above)
Comment (UselessToucan) [reply osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: `config.GetBindable<Size>(FrameworkSetting.SizeFullscreen)` call

Comment (FreezyLemon) [reply osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: Oh, alright. You could always use `config.BindWith` instead, if you want to.

File: osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs [hunk 6]
```diff
@@ -126,7 +126,7 @@ private void load(FrameworkConfigManager config, OsuGameBase game)
 
         private List<KeyValuePair<string, int>> getResolutions()
         {
-            var availableDisplayResolutions = (game.Window as DesktopGameWindow)?.AvailableDisplayResolutions
+            var availableDisplayResolutions = (game.Window as DesktopGameWindow)?.GetCurrentDisplay().AvailableResolutions
```
Comment (FreezyLemon) [inline osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs]: `GetCurrentDisplay` is a member of `GameWindow` on purpose - no need to cast :)
Also, you should call `.Distinct()` here (it got lost with the `AvailableDisplayResolutions` property)

**Review by peppy — APPROVED** [review]

**Pull request merged**